<a href="https://colab.research.google.com/github/VitorFRodrigues/Data-Science-Bootcamp/blob/main/Proj05_final/Notebook/Bootcamp_Proj_05.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Projeto Módulo 5 do Bootcamp Data Science 3 - Criação de modelo preditivo para decisão de paciente se encontra apto ou não para ingressar na UTI
---
Autor: Vitor Rodrigues

e-mail: vitorfbaiano@gmail.com

---

## 1.Introdução

Explicar um pouco sobre o problema proposto, a quantidade de dados coletado pelo Sirio Libanes, como e onde estes dados são encontrados.

## 2.Objetivos

Considerando o que foi explanado na introdução, este projeto tem como objetivo a criação de um modelo decisório para ingresso de pacientes na UTI dados certas variárieis. Para isto, serão realizados os seguintes passos:

- Importação dos dados do Sirio Libanês;
- Realização de limpeza dos dados;
- Criação de 5 modelos preditivos;
- Alteração de hiperparametros para atingir melhor acurácia em cada modelo;
- Remover informações correlacionadas que dificultem o poder decisório de cada modelo;
- Aplicar os modelos em novo banco de dados para buscar melhor acurácia;
- Apurar modelo preditivo e salva-lo;

Tipos principais de modelos preditivos:


*   Classificação: escolha booleana (Verdadeiro ou Falso)
*   Regressão: previsão dentro de um intervalo ]-inf,+inf[ 



## 3.Preparação dos dados

Falar do Sirio Libanes e como este levantou estes dados. Informar a metodologia das 

Regras dos dados recebidos
- Alguns itens NaN;
- 384 pacientes (vide numero máximo no PACIENTE_VISIT_IDENTIFIER)
- Cada paciente separado em 5 janelas de tempo (vide coluna Window)

In [1]:
import pandas as pd

Sirio_Libanes_URL = 'https://github.com/VitorFRodrigues/Data-Science-Bootcamp/blob/main/Proj05_final/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true'

In [2]:
dados = pd.read_excel(Sirio_Libanes_URL)
dados

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2                  NaN                     NaN   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                                  NaN                              NaN   
3                                  NaN                              NaN   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863                   

dar informações dos dados, numero de linhas e colunas, o que as colunas representam (falar dos 54 tipos), falar que o paciente está subdividido em 5 janelas. Informar a metodologia da janela.

In [3]:
dados[['PATIENT_VISIT_IDENTIFIER', 'WINDOW']].head()

PATIENT_VISIT_IDENTIFIER    WINDOW
0                         0       0-2
1                         0       2-4
2                         0       4-6
3                         0      6-12
4                         0  ABOVE_12

puxar o dados.describe(). Informar que este descreve todas as colunas que são do tipo inteira ou categórica

In [4]:
dados.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  ...  \
count         1920.000000         1920.000000  1920.000000  ...   
mean             0.128125            0.046875     0.213021  ...   
std              0.334316            0.211426     0.409549  ...   
min              0.000000            0.000000     0.000000  ...   
25%              0.000000            0.000000     0.000000  ...   
50%              0.000000            0.000000     0.000000  ...   
75%              0.000000            0.000000     0.000000  ...   
max              1.000000            1.000000     1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1177.000000       1231.000000             1239.000000   
mean               -0.703683         -0.770338               -0.887196   
std                 0.482097          0.319001                0.296147   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -0.976190               -0.979798   
75%                -0.647059         -0.595238               -0.878788   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1240.000000                      1240.000000   
mean                          -0.786997                        -0.715950   
std                            0.324754                         0.419103   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -0.984944   
75%                           -0.645482                        -0.522176   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1240.000000                1177.000000           1231.000000   
mean             -0.817800                  -0.719147             -0.771327   
std               0.270217                   0.446600              0.317694   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -0.989822                  -1.000000             -0.975924   
75%              -0.662529                  -0.634409             -0.594677   
max               1.000000                   1.000000              

Mostrar a anomalia q existem apenas 229 colunas do tipo inteiras ou categóricas. Vamos informar quais são as colunas não categorizadas

In [5]:
for coluna in dados.columns:
  if coluna not in dados.describe().columns:
    print(coluna)

AGE_PERCENTIL
WINDOW


In [6]:
dados[['AGE_PERCENTIL', 'WINDOW']]

AGE_PERCENTIL    WINDOW
0             60th       0-2
1             60th       2-4
2             60th       4-6
3             60th      6-12
4             60th  ABOVE_12
...            ...       ...
1920          50th       0-2
1921          50th       2-4
1922          50th       4-6
1923          50th      6-12
1924          50th  ABOVE_12

[1925 rows x 2 columns]

Informar que a janela window será agrupada posteriormente e por isso não será categorizada. Já a coluna AGE_PERCENTIL indica a.... e será categorizada abaixo:

Para categorizar, existem algumas metodologias:
1. Find and Replace;
2. Label Encoding;
3. One Hot Encoding
4. OriginalEncoder (do Scikit-Learn)
5. OneHotEncoder (do Scikit-Learn)

Será usada a primeira. Para ver como funciona as demais metodologias de categorização, vide [link](https://github.com/VitorFRodrigues/Data-Science-Bootcamp/blob/main/Proj05_final/Notebook/Bootcamp_Proj_05-Metodologias_de_Categorizacao.ipynb).


In [7]:
dados_categorizar = dados['AGE_PERCENTIL']
dados_categorizar.unique()

array(['60th', '90th', '10th', '40th', '70th', '20th', '50th', '80th',
       '30th', 'Above 90th'], dtype=object)

In [8]:
dados_categorizar.value_counts()

20th          215
10th          205
30th          205
40th          200
70th          195
50th          190
80th          190
60th          185
Above 90th    185
90th          155
Name: AGE_PERCENTIL, dtype: int64

In [9]:
substituir = {"60th": 60, 
              "10th": 10, 
              "40th": 40, 
              "70th": 70, 
              "20th": 20, 
              "50th": 50, 
              "80th": 80, 
              "30th": 30,
              "90th": 90, 
              "Above 90th": 100}
dados_categorizado_find_replace = dados_categorizar.replace(substituir)
dados_categorizado_find_replace

0       60
1       60
2       60
3       60
4       60
        ..
1920    50
1921    50
1922    50
1923    50
1924    50
Name: AGE_PERCENTIL, Length: 1925, dtype: int64

A partir deste momento, estamos modificando os dados originais, sendo assim estes serão adicionados em novo dataframe dados_limpos.

In [10]:
dados_limpos = dados
dados_limpos['AGE_PERCENTIL'] = dados_categorizado_find_replace

In [11]:
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL       GENDER  \
count               1925.000000  1925.000000    1925.000000  1925.000000   
mean                 192.000000     0.467532      53.194805     0.368831   
std                  111.168431     0.499074      28.673479     0.482613   
min                    0.000000     0.000000      10.000000     0.000000   
25%                   96.000000     0.000000      30.000000     0.000000   
50%                  192.000000     0.000000      50.000000     0.000000   
75%                  288.000000     1.000000      80.000000     1.000000   
max                  384.000000     1.000000     100.000000     1.000000   

       DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
count         1920.000000         1920.000000         1920.000000   
mean             0.108333            0.028125            0.097917   
std              0.310882            0.165373            0.297279   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
count         1920.000000         1920.000000         1920.000000  ...   
mean             0.019792            0.128125            0.046875  ...   
std              0.139320            0.334316            0.211426  ...   
min              0.000000            0.000000            0.000000  ...   
25%              0.000000            0.000000            0.000000  ...   
50%              0.000000            0.000000            0.000000  ...   
75%              0.000000            0.000000            0.000000  ...   
max              1.000000            1.000000            1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1177.000000       1231.000000             1239.000000   
mean               -0.703683         -0.770338               -0.887196   
std                 0.482097          0.319001                0.296147   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -0.976190               -0.979798   
75%                -0.647059         -0.595238               -0.878788   
max                 1.000000          1.000000                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1240.000000                      1240.000000   
mean                          -0.786997                        -0.715950   
std                            0.324754                         0.419103   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -0.984944   
75%                           -0.645482                        -0.522176   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1240.000000                1177.000000           1231.000000   
mean             -0.817800                  -0.719147             -0.771327   
std               0.270217                   0.446600              0.317694   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -0.989822                  -1.000000             -0.975924   
75%              -0.662529                  -0.634409             -0.594677   
max               1.000000                   1.00

Agora temos todas as variaveis categoricas necessárias para aplicação do modelo. Agora podemos tratar os dados. Como já foi explicado, temos 384 pacientes subdivididos em 5 janelas de tempo.

"Se a pessoa ta na UTI eu não posso usar aquela informação. Pára pra pensar em um caso específico, se a pessoa entra no Hospital e já está na UTI, nas 2 primeiras horas já entra na UTI, a gente não pode usar aquele paciente, aquele paciente vai embora por que na primeira leitura ele já está na UTI."

Explicar que não faz sentido usar no nosso projeto aqueles paciente que já se encontram na UTI nas 2 primeiras horas, sendo assim, o código abaixo vai retirar os pacientes que estão na UTI nas 2 primeiras horas

In [12]:
a_remover = dados_limpos.query("WINDOW=='0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER'].values
dados_limpos = dados_limpos.query('PATIENT_VISIT_IDENTIFIER not in @a_remover')
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL       GENDER  \
count               1765.000000  1765.000000    1765.000000  1765.000000   
mean                 192.835694     0.456091      52.662890     0.379603   
std                  110.481302     0.498209      28.882085     0.485426   
min                    0.000000     0.000000      10.000000     0.000000   
25%                   98.000000     0.000000      30.000000     0.000000   
50%                  192.000000     0.000000      50.000000     0.000000   
75%                  289.000000     1.000000      80.000000     1.000000   
max                  384.000000     1.000000     100.000000     1.000000   

       DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
count          1760.00000         1760.000000         1760.000000   
mean              0.11250            0.026136            0.101705   
std               0.31607            0.159586            0.302345   
min               0.00000            0.000000            0.000000   
25%               0.00000            0.000000            0.000000   
50%               0.00000            0.000000            0.000000   
75%               0.00000            0.000000            0.000000   
max               1.00000            1.000000            1.000000   

       DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
count         1760.000000         1760.000000         1760.000000  ...   
mean             0.021591            0.132955            0.049432  ...   
std              0.145385            0.339622            0.216829  ...   
min              0.000000            0.000000            0.000000  ...   
25%              0.000000            0.000000            0.000000  ...   
50%              0.000000            0.000000            0.000000  ...   
75%              0.000000            0.000000            0.000000  ...   
max              1.000000            1.000000            1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count            1053.000000       1103.000000             1111.000000   
mean               -0.721133         -0.775375               -0.892071   
std                 0.466468          0.308266                0.288869   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -0.976190               -0.979798   
75%                -0.647059         -0.595238               -0.878788   
max                 1.000000          0.333333                1.000000   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                       1108.000000                      1108.000000   
mean                          -0.796061                        -0.729520   
std                            0.311507                         0.402389   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -0.647275                        -0.541827   
max                            1.000000                         1.000000   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count          1112.000000                1053.000000           1103.000000   
mean             -0.819947                  -0.733648             -0.776424   
std               0.267913                   0.433956              0.306903   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -0.976451   
75%              -0.664947                  -0.681582             -0.595796   
max               1.000000                   1.00

Pode ser visto que a contagem anterior era 1.925 e agora estamos com 1.765. A diferença é igual a 160 e como temos 5 PATIENT_VISIT_IDENTIFIER para cada paciente, subdividimos o valor por 5 e descobrimos que 32 pacientes foram removidos da lista.

Próximo ponto será manipulação de valores nulos.

In [15]:
is_NaN = dados_limpos.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = dados_limpos[row_has_NaN]
rows_with_NaN.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  AGE_PERCENTIL       GENDER  \
count               1290.000000  1290.000000    1290.000000  1290.000000   
mean                 192.041085     0.456589      52.627907     0.380620   
std                  110.573927     0.498305      28.725060     0.485728   
min                    0.000000     0.000000      10.000000     0.000000   
25%                   97.250000     0.000000      30.000000     0.000000   
50%                  190.500000     0.000000      50.000000     0.000000   
75%                  288.750000     1.000000      80.000000     1.000000   
max                  384.000000     1.000000     100.000000     1.000000   

       DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
count         1285.000000         1285.000000         1285.000000   
mean             0.098054            0.021012            0.091829   
std              0.297504            0.143479            0.288897   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
count         1285.000000         1285.000000         1285.000000  ...   
mean             0.018677            0.124514            0.042802  ...   
std              0.135434            0.330295            0.202488  ...   
min              0.000000            0.000000            0.000000  ...   
25%              0.000000            0.000000            0.000000  ...   
50%              0.000000            0.000000            0.000000  ...   
75%              0.000000            0.000000            0.000000  ...   
max              1.000000            1.000000            1.000000  ...   

       RESPIRATORY_RATE_DIFF  TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
count             578.000000        628.000000              636.000000   
mean               -0.950132         -0.951054               -0.981196   
std                 0.138757          0.102627                0.041803   
min                -1.000000         -1.000000               -1.000000   
25%                -1.000000         -1.000000               -1.000000   
50%                -1.000000         -1.000000               -1.000000   
75%                -1.000000         -0.952381               -0.979798   
max                 0.294118         -0.238095               -0.474747   

       BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
count                        633.000000                       633.000000   
mean                          -0.956410                        -0.943864   
std                            0.106839                         0.136183   
min                           -1.000000                        -1.000000   
25%                           -1.000000                        -1.000000   
50%                           -1.000000                        -1.000000   
75%                           -0.986874                        -0.976560   
max                            0.027425                         0.092501   

       HEART_RATE_DIFF_REL  RESPIRATORY_RATE_DIFF_REL  TEMPERATURE_DIFF_REL  \
count           637.000000                 578.000000            628.000000   
mean             -0.965528                  -0.953988             -0.951721   
std               0.082787                   0.125805              0.100860   
min              -1.000000                  -1.000000             -1.000000   
25%              -1.000000                  -1.000000             -1.000000   
50%              -1.000000                  -1.000000             -1.000000   
75%              -0.986164                  -1.000000             -0.952248   
max              -0.417142                   0.34

Várias linhas deste DataFrame possui pelo menos uma coluna contendo NaN (Not-a-Number). O código anterior mostra que das 1765 linhas, 1290 possuem um valor nulo em alguma coluna, representando mais de 70% dos dados. Retirar todas estas linhas não faz sentido e nem auxiliará na criação do modelo preditivo. 

Sendo assim, vamos preencher estes valores nulos com informações. Mas não serão qualquer números, vamos separar cada paciente em suas respectivas janelas de tempo e preencher os valores nulos com o mesmo número indicado na linha anterior ou posterior.

 Para melhor exemplificar o que estou falando, vide este [link](https://github.com/VitorFRodrigues/Data-Science-Bootcamp/blob/main/Proj05_final/Notebook/Previs%C3%A3o_de_Incrementos.ipynb).